### DISCUSSION TITLE
[v5] How to add a new Mail driver? (PostmarkApp)

### DISCUSSION THREAD

**Discussion by ivan-prats** [adonisjs/core#2053] [General]
Hey guys! First time posting so please forgive me if i'm not following some guideling.

I love [PostmarkApp](https://postmarkapp.com/) and I also love the Adonis v5 Mailer package. So I wouldn't mind looking into adapting a driver for it. 

I did found [this docs section](https://adonisjs.com/docs/3.2/mail#_adding_new_drivers) on the legacy 3.2 Adonis version on how to extend drivers, even with an example with PostmarkApp! Introducing the [nodemailer-postmark-transport package](https://github.com/killmenot/nodemailer-postmark-transport).

But being docs of a legacy release: I did prefer to ask first about it before trying to more or less follow it. Is there any docs with more updated info? Or does that doc section remain more or less true nowadays? 

I could also try piecing that doc section togheter with the current Mail Driver syntax ([of the Mailgun Driver for example?](https://github.com/adonisjs/mail/blob/4bdc4619bd91fc6824da8383eb3e5b60642fc8b6/src/Drivers/Mailgun.ts))

Any helpful hint on where to start looking into it would be great. And if I manage to do it: I will be sure to post about it online and share it. Thanks!

**Comment by thetutlage** [comment]
Hello 👋

Yup would love to have a driver for postmark. I am sure if they have an endpoint that accepts a mime payload. If they do, then it can be an exact copy of the mailgun driver.

**Reply by ivan-prats** [reply]
Mmmm I think they don't accept a MIME string payload per-se:
![image](https://user-images.githubusercontent.com/63517766/102373525-9326db00-3fc0-11eb-98d7-e74278a31ace.png)
[Source](https://postmarkapp.com/migration-guides/mailgun#differences-from-mailgun)

But I guess that as long as we can comunicate with their API endpoint with MIME compoments it's fine?

I think the biggest chunk of work would be initiating the PostmarkTransport. There is [this package](https://github.com/killmenot/nodemailer-postmark-transport) that literally comes with a [PostmarkTransport](https://github.com/killmenot/nodemailer-postmark-transport/blob/master/lib/postmark-transport.js) class already prepared.

Buuut that package uses the [postmark official package](https://github.com/wildbit/postmark.js) to make the API calls. 

I have two big questions looking more into re-fractoring what I will do into a future PR:
1. Would you @thetutlage prefer to avoid adding any additional packages, and communicate with Postmark's API through HTTP requests? (with "got" for example)
2. Looking into writing the PostmarkTransport class: what are the requirements for a adonisjs/mail Driver & Transport? 
[I can see](https://github.com/adonisjs/mail/blob/4bdc4619bd91fc6824da8383eb3e5b60642fc8b6/src/Drivers/Mailgun.ts#L35) that the Transport needs to be passed on into the `nodemailer.createTransport` method. But is there any additional requirements to integrate correctly into AdonisJS ?

**Reply by ivan-prats** [reply]
I've been toying with this, but I'm finding it really hard to know where to extend the `MailManager` with the new PostmarkDriver in Adonis v5.

Looking through the code base I *guess* I have to do something like:
```typescript
MailManager.extend('postmark', () => new PostmarkDriver({}, Logger))
```

But I'm obviously doing something wrong extending the Driver because I get this message:
```bash
# This is in the repl. After creating a "Testing" Mailer than only sends a "<p>hello</p>" message
> loadMailers()
> await new mailers.Testing().send()
Uncaught Error: "postmark" driver is not supported by "MailManager"
```

Where should I run the first code to extend the MailDriver? In the `./start` folder? As a command and run it somewhere? Maybe I have to do something else on top of extending it?

**Reply by thetutlage** [reply]
I created a sample repo here showing how to extend the Mailer to add new drivers https://github.com/thetutlage/postmark-driver-app

**Reply by thetutlage** [reply]
Regarding the packages to use. I am okay using the official client of PostMark https://github.com/wildbit/postmark.js and build on top of it. Couple of things to make sure

- We are able to attach files
- We are able to embed files using CID
- We are able to define the content for plain text and the Apple watch view

**Reply by ivan-prats** [reply]
Files attachements and embeding files using CID will be fine!

Plain text content is fine for sure: 
![image](https://user-images.githubusercontent.com/63517766/102803408-7a4f6880-43b8-11eb-9ea3-4c5bf94284c2.png)


But I'm finding it hard to find documentation / requirements about sending emails taking into account that it's an Apple Watch? Does it work like a normal Plain Text email? I can't find anything for apple-watch-specific-formats

**Reply by thetutlage** [reply]
Yup, looking at their API docs, it looks like there is no support for watch html and that is okay, we can write a note about it in the docs. The main idea was to have 100% support for what can be achieved with other drivers and watch html is not that of a big deal.

Generally it is represented as follows inside a MIME string. https://www.litmus.com/blog/how-to-send-hidden-version-email-apple-watch/

![mime-string](https://user-images.githubusercontent.com/1706381/102849286-d8be2a80-443c-11eb-94ad-a88ce0d6c3f1.png)
